freqtrade/data/converter.py

```diff
@@ -121,6 +121,45 @@ def ohlcv_fill_up_missing_data(dataframe: DataFrame, timeframe: str, pair: str)
             logger.debug(message)
     return df
 
+def reduce_mem_usage(pair: str, df: DataFrame) -> DataFrame:
+    """ iterate through all the columns of a dataframe and modify the data type
+        to reduce memory usage.        
+    """
+    # start_mem = df.memory_usage().sum() / 1024**2
+    # logger.info(f"Memory usage of dataframe for {pair} is {start_mem:.2f} MB")
+   
+    for col in df.columns[1:]:
+        col_type = df[col].dtype
+        
+        if col_type != object:
+            c_min = df[col].min()
+            c_max = df[col].max()
+            if str(col_type)[:3] == "int":
+                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
```

Comment:
I do think this method goes too far in it's current form.

the method introduced in #7701 (`reduce_dataframe_footprint()`) does something similar (down to int32) - but in a more performant way.
it's also skipping regular ohlcv columns to avoid problematic conversions on core data.

on bigger (wider) dataframes, we've observed looping columns to change their type to have a long runtime (observed up to >100s) - which obviously defeats the purpose, as we'd be trading memory for runtime.

Reply:
this method is aimed for backtest and hyperopt, so the runtime isn't a big issue. Also since we already have all the data, slimming down the type below 32 isn't an issue and really helps reducing more memory, which is really a good thing

Reply:
i disagree with this. Runtime is always an issue.

depending on the mode - trimming may happen for every hyperopt epoch.
this means in the sample i mentioned above - we'd spend 1m per epoch (!) on adjusting datatypes - instead of actually running hyperopt.

Reply:
I hope the "1 extra min per epoch" isn't something coming out of thin air.

I done a stress test. Use this strat https://github.com/stash86/MultiMA_TSL/blob/main/user_data/strategies/MultiMA_TSL/MultiMA_TSL5.py
and this pairlist https://github.com/stash86/freqtrade/blob/freqtrade-develop/user_data/other_configs/config-static-busd.json

run a backtest for 20210901-20220901

and by stress test, I really take the most out of the memory. And the result?

current develop version
```
2022-11-25 10:41:42,376 - freqtrade - INFO - freqtrade docker-2022.11.dev-c963fd72
2022-11-25 11:25:39,396 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2021-09-01 00:00:00 up to 2022-09-01 00:00:00 (365 days).
```
which means it took 43 mins 57s to load stuffs and calculate all indicators


my proposed code
```
2022-11-25 09:33:01,267 - freqtrade - INFO - freqtrade 2022.11.dev
2022-11-25 10:17:16,305 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2021-09-01 00:00:00 up to 2022-09-01 00:00:00 (365 days).
```
took 44 mins 16s

only 19s extra for a massive backtest.

Reply:
How many columns does it result in the end? (how many are returned after the last populate_* call)?

Can you try it with a the already existing version / call (`reduce_dataframe_footprint()`), too?

> I hope the "1 extra min per epoch" isn't something coming out of thin air.

It's been tested for FreqAI usage - with a huge amount (not sure how many, but > 1000) columns.

Reply:
So I reduced the timerange to 3 months (don't want to wait for another 45 mins)

modified the trim_dataframes to do 3 separate trimming actions
```
def trim_dataframes(preprocessed: Dict[str, DataFrame], timerange,
                    startup_candles: int) -> Dict[str, DataFrame]:
    """
    Trim startup period from analyzed dataframes
    :param preprocessed: Dict of pair: dataframe
    :param timerange: timerange (use start and end date if available)
    :param startup_candles: Startup-candles that should be removed
    :return: Dict of trimmed dataframes
    """
    processed: Dict[str, DataFrame] = {}
    processed1: Dict[str, DataFrame] = {}
    processed2: Dict[str, DataFrame] = {}

    memory_used = 0
    tik = time.perf_counter()
    for pair, df in preprocessed.items():
        trimed_df = trim_dataframe(df, timerange, startup_candles=startup_candles)
        if not trimed_df.empty:
            # start_mem = trimed_df.memory_usage().sum() / 1024**2
            # logger.info(f"Memory usage of dataframe for {pair} before reduced is {start_mem:.2f} MB")
            # trimed_df = reduce_mem_usage(pair, trimed_df)
            # trimed_df = reduce_dataframe_footprint(trimed_df)
            # end_mem = trimed_df.memory_usage().sum() / 1024**2
            # logger.info(f"Memory usage of dataframe for {pair} after reduced is {end_mem:.2f} MB")
            processed[pair] = trimed_df
            memory_used += trimed_df.memory_usage().sum() / 1024**2
        else:
            logger.warning(f'{pair} has no data left after adjusting for startup candles, '
                           f'skipping.')
    tok = time.perf_counter()
    logger.info(f"Memory usage of original processed is {memory_used:.2f} MB")
    logger.info(f"No-optimization took {tok - tik:0.4f} seconds to trim dataframes")

    memory_used = 0
    tik = time.perf_counter()
    for pair, df in preprocessed.items():
        trimed_df = trim_dataframe(df, timerange, startup_candles=startup_candles)
        if not trimed_df.empty:
            trimed_df = reduce_dataframe_footprint(trimed_df)
            memory_used += trimed_df.memory_usage().sum() / 1024**2
            processed1[pair] = trimed_df
        else:
            logger.warning(f'{pair} has no data left after adjusting for startup candles, '
                           f'skipping.')

    tok = time.perf_counter()
    logger.info(f"Memory usage of loose optimization processed is {memory_used:.2f} MB")
    logger.info(f"Loose optimization took {tok - tik:0.4f} seconds to trim dataframes")

    memory_used = 0
    tik = time.perf_counter()
    for pair, df in preprocessed.items():
        trimed_df = trim_dataframe(df, timerange, startup_candles=startup_candles)
        if not trimed_df.empty:
            trimed_df = reduce_mem_usage(pair, trimed_df)
            memory_used += trimed_df.memory_usage().sum() / 1024**2
            processed2[pair] = trimed_df
        else:
            logger.warning(f'{pair} has no data left after adjusting for startup candles, '
                           f'skipping.')

    tok = time.perf_counter()
    logger.info(f"Memory usage of aggresive optimization processed is {memory_used:.2f} MB")
    logger.info(f"Aggresive optimization took {tok - tik:0.4f} seconds to trim dataframes")

    return processed
```

Since your concern is how much extra times needed to trim the dataframes' type, so I only focus to check the timing on those parts. And the result
```
2022-11-25 15:06:11,699 - freqtrade.data.converter - INFO - Memory usage of original processed is 1355.09 MB
2022-11-25 15:06:11,699 - freqtrade.data.converter - INFO - No-optimization took 1.4424 seconds to trim dataframes
2022-11-25 15:06:14,317 - freqtrade.data.converter - INFO - Memory usage of loose optimization processed is 908.47 MB
2022-11-25 15:06:14,317 - freqtrade.data.converter - INFO - Loose optimization took 2.6179 seconds to trim dataframes
2022-11-25 15:06:20,132 - freqtrade.data.converter - INFO - Memory usage of aggresive optimization processed is 541.84 MB
2022-11-25 15:06:20,132 - freqtrade.data.converter - INFO - Aggresive optimization took 5.8147 seconds to trim dataframes
```

Loose optimize =  current freqai reduce_footprint function
Aggresive optimize = this PR

Reply:
For a 6-month backtest
```
2022-11-26 05:50:49,299 - freqtrade.data.converter - INFO - Memory usage of original processed is 2649.22 MB
2022-11-26 05:50:49,300 - freqtrade.data.converter - INFO - No-optimization took 2.1248 seconds to trim dataframes
2022-11-26 05:50:52,926 - freqtrade.data.converter - INFO - Memory usage of loose optimization processed is 1776.07 MB
2022-11-26 05:50:52,926 - freqtrade.data.converter - INFO - Loose optimization took 3.6262 seconds to trim dataframes
2022-11-26 05:51:02,188 - freqtrade.data.converter - INFO - Memory usage of aggresive optimization processed is 1059.35 MB
2022-11-26 05:51:02,188 - freqtrade.data.converter - INFO - Aggresive optimization took 9.2616 seconds to trim dataframes
```